File: WordPattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {

        //pattern->string
        HashMap<Character, String> map1 = new HashMap<>();
        //string->pattern
        HashMap<String, Character> map2 = new HashMap<>();

        String[] words = s.split(" ");
        if(words.length != pattern.length())
            return false;

        for(int i = 0; i < words.length; i++) {

            char ch = pattern.charAt(i);
            String str = words[i];

            if(map1.containsKey(ch)) {

                if(!map1.get(ch).equals(str))
                    return false;
            }
            else
                map1.put(ch, str);
            
            if(map2.containsKey(str)) {

                if(map2.get(str) != ch) 
                    return false;
            }
            else
                map2.put(str, ch);
        }
        
        return true;
    }
}

File: WordPattern.cs
public class Solution {
    // Time Complexity: O(n), where n is the number of characters in the pattern (and words in the string).
    // Space Complexity: O(n), where n is the number of characters in the pattern (and words in the string).
    
    public bool WordPattern(string pattern, string s) {
    
        string[] words = s.Split(' ');

        if (pattern.Length != words.Length) return false;

        Dictionary<char, string> charToWord = new Dictionary<char, string>();
        Dictionary<string, char> wordToChar = new Dictionary<string, char>();

        for (int i = 0; i < pattern.Length; i++)
        {
            char charPattern = pattern[i];  
            string word = words[i];         

            if (charToWord.ContainsKey(charPattern))
            {
                if (charToWord[charPattern] != word)
                {
                    return false; 
                }
            }
            else
            {
                charToWord[charPattern] = word;
            }

            if (wordToChar.ContainsKey(word))
            {
                if (wordToChar[word] != charPattern)
                {
                    return false;  
                }
            }
            else
            {
                wordToChar[word] = charPattern;
            }
        }

        return true;
    }
}
